torch/distributed/_tensor/tp_conv.py

```diff
@@ -0,0 +1,230 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+# implement matrix related ops for distributed tensor
+from typing import cast, Dict, Tuple
+
+import torch
+import torch.distributed as dist
+
+aten = torch.ops.aten
+
+
+def _requires_data_exchange(padding):
+    # TODO: whether there requires data exchange is currently determined by padding
+    return padding[1] != 0
+
+
+def _is_supported(input_size, kernel_size, stride, padding, dilation):
+    if dilation[1] != 1:
+        raise RuntimeError("Dilation must be 1 for tensor parallel convolution.")
+    if padding[1] != 0:
+        if stride[1] != 1:
+            raise RuntimeError(
+                "Stride must be 1 when there is padding for tensor parallel convolution."
+            )
+        if kernel_size[3] // 2 > input_size[3]:
+            raise RuntimeError(
+                "kernel_size[3] // 2 should be less than or equal to input_size[3] for tensor parallel convolution."
+            )
+    else:
+        if not (input_size[3] % stride[1] == 0 and stride[1] == kernel_size[3]):
+            raise RuntimeError(
+                "It requires that input_size[3] is divisible by stride[1] and stride[1] equals kernel_size[3] "
+                "when there is padding for tensor parallel convolution."
+            )
+    return True
+
+
+def tp_convolution(
+    op_call: torch._ops.OpOverload,
+    *local_tensor_args: Tuple[object, ...],
+    **local_tensor_kwargs: Dict[str, object],
+) -> object:
+    assert op_call == aten.convolution.default
+    assert len(local_tensor_args) == 9
+
+    rank = dist.get_rank()
+    size = dist.get_world_size()
+    in_tensor = local_tensor_args[0]
+    weight = local_tensor_args[1]
+    bias = local_tensor_args[2]
+    stride, padding, dilation = local_tensor_args[3:6]
+
+    assert _is_supported(in_tensor.shape, weight.shape, stride, padding, dilation)
+
+    if not _requires_data_exchange(padding):
+        local_results = op_call(*local_tensor_args, **local_tensor_kwargs)
+        return local_results
+    else:
+        # step 0 compute the overlap pixels of the input tensor
+        d = weight.shape[3] - 1
```

Comment:
I am trying to see if we should make these batch send/recv calls be a type of redistribute/resharding, so instead of having a custom convolution op implemented, and special case in torch_dispatch, we should try to see if:
- for the case of input data exchange, we can make this input send/recvs be a data order permutation (i.e. sth like mesh [0, 1, 2, 3] to mesh [1, 2, 3, 0]).
- I see duplicate comms happen in the backward op too, we should merge these two into a common function.

Reply:
I have merged communications as a function for conv fwd/bwd.

Regarding replacing batch send/recv calls with redistribute/resharding, we can keep the discussion, and I think it will not happen in this PR.